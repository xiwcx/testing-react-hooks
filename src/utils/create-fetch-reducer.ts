type ReducerState<T> = {
  loading: boolean;
  error: string | null;
  data: T[];
}

type ReducerAction<T> =
  | {type: 'error', error: string}
  | {type: 'loading'}
  | { data: T[]; type: 'success'}

export const createFetchReducer = <T>() => (
  prevState: ReducerState<T>,
  action: ReducerAction<T>
): ReducerState<T> => {
  switch (action.type) {
    case 'error':
      return {
        loading: false,
        data: prevState.data,
        error: action.error,
      }
    case 'loading':
      return {
        loading: true,
        data: prevState.data,
        error: null,
      }
    case 'success':
      return {
        loading: false,
        data: action.data,
        error: null
      }
    default:
      return prevState;
  }
}
