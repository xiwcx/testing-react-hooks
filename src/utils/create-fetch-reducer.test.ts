import { renderHook, act } from '@testing-library/react-hooks';
import { useReducer } from 'react';
import { createFetchReducer } from './create-fetch-reducer';

type Item = {
  id: string
}

const items: Item[] = [
  { id: '1'},
  { id: '2'},
  { id: '3'},
]

const initialState = {
  error: null,
  loading: true,
  data: [],
}

test('should return correct state after dispatch', () => {
  const { result } = renderHook(() => useReducer(createFetchReducer<Item>(), initialState))

  const [state, dispatch] = result.current;

  expect(state).toMatchSnapshot();

  act(() => dispatch({type: 'success', data: items}))

  expect(result.current[0]).toMatchSnapshot();

  act(() => dispatch({type: 'error', error: 'oh no!'}))

  expect(result.current[0]).toMatchSnapshot();

  act(() => dispatch({type: 'loading'}))

  expect(result.current[0]).toMatchSnapshot();
})
