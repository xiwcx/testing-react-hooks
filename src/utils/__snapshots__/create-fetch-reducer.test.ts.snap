// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct state after dispatch 1`] = `
Object {
  "data": Array [],
  "error": null,
  "loading": true,
}
`;

exports[`should return correct state after dispatch 2`] = `
Object {
  "data": Array [
    Object {
      "id": "1",
    },
    Object {
      "id": "2",
    },
    Object {
      "id": "3",
    },
  ],
  "error": null,
  "loading": false,
}
`;

exports[`should return correct state after dispatch 3`] = `
Object {
  "data": Array [
    Object {
      "id": "1",
    },
    Object {
      "id": "2",
    },
    Object {
      "id": "3",
    },
  ],
  "error": "oh no!",
  "loading": false,
}
`;

exports[`should return correct state after dispatch 4`] = `
Object {
  "data": Array [
    Object {
      "id": "1",
    },
    Object {
      "id": "2",
    },
    Object {
      "id": "3",
    },
  ],
  "error": null,
  "loading": true,
}
`;
